tensorrt_llm/_torch/auto_deploy/transformations/library/sharding.py

```diff
@@ -319,3 +319,126 @@ def column_row_shard(gm: GraphModule, rank: int, world_size: int) -> GraphModule
     gm = canonicalize_graph(gm)
     ad_logger.debug("After sharding: " + str(gm))
     return gm
+
+
+def bmm_shard(gm: GraphModule, rank: int, world_size: int) -> GraphModule:
+    """A transformation to apply sharding to batched matrix multiplications in the graph.
+
+    We'll shard the BMM nodes by slicing the batch dimension of input tensors into world_size number of slices.
+    After sharding each BMM node, we'll insert an all_gather node to gather the results across the different devices.
+    This transformation handles any combination of tensor types for both inputs to the BMM operation.
+
+    We'll also assume that the inputs to BMM are broadcasted across the devices already.
+    """
+    ad_logger.info("Sharding graph for BMM")
+    ad_logger.debug("Before sharding graph: " + str(gm))
+
+    if world_size < 2:
+        ad_logger.info("Skipping sharding for single device")
+        return gm
+
+    assert isinstance(gm, GraphModule), "Expecting GraphModule"
+
+    def handle_tensor(
```

Comment:
@meenchen, here is an example of bmm sharding where either tensor can be a weight or activation tensor